utildb/gen_deleted_accounts.go

```diff
@@ -59,100 +67,284 @@ func readAccounts(ch chan proxy.ContractLiveliness, deleteHistory *map[common.Ad
+	workerInputChannels := taskFeeder(&wg, cfg, iter, lastBlock, stopChan, log)
+
+	// prepare workers to process transactions
+	workerOutputChannels := launchWorkers(&wg, cfg, workerInputChannels, processor, stopChan, errChan, log)
+
+	// collect results from workers and orders them
+	orderedResults := resultCollector(&wg, cfg, workerOutputChannels, stopChan)
+
+	// process ordered txLivelinessResults
+	resolveDeletionsAndResurrections(ddb, orderedResults, stopChan, errChan)
 
-	for iter.Next() {
-		tx := iter.Value()
-		if tx.Block > lastBlock {
-			break
+	// wait until feeder, workers and collector are done
+	wg.Wait()
+
+	// notify error handler to stop
+	close(errChan)
+
+	utils.StopCPUProfile(cfg)
+
+	// retrieve encounteredErrors from error handler
+	err = <-encounteredErrors
+	return err
+}
+
+// resolveDeletionsAndResurrections reads txLivelinessResults and resolves deletions and resurrections.
+func resolveDeletionsAndResurrections(ddb *substate.DestroyedAccountDB, orderedResults chan txLivelinessResult, stopChan chan struct{}, errChan chan error) {
+	var deleteHistory = make(map[common.Address]bool)
+	defer func() {
+		// explicitly set to nil to release memory as soon as possible
+		deleteHistory = nil
+	}()
+
+	for {
+		select {
+		case <-stopChan:
+			return
+		case contract, ok := <-orderedResults:
+			{
+				if !ok {
+					return
+				}
+				des, res := readAccounts(contract.liveliness, &deleteHistory)
+				if len(des)+len(res) > 0 {
+					err := ddb.SetDestroyedAccounts(contract.tx.Block, contract.tx.Transaction, des, res)
+					if err != nil {
+						errChan <- err
+						return
+					}
+				}
+			}
 		}
+	}
+}
+
+// resultCollector collects results from workers in round-robin fashion and sends them to a single channel.
+func resultCollector(wg *sync.WaitGroup, cfg *utils.Config, workerOutputChannels map[int]chan txLivelinessResult, stopChan chan struct{}) chan txLivelinessResult {
+	orderedResults := make(chan txLivelinessResult, cfg.Workers)
+	wg.Add(1)
+	go func() {
+		defer close(orderedResults)
+		defer wg.Done()
 
-		if tx.Transaction < utils.PseudoTx {
-			err = genDeletedAccountsTask(tx, processor, ddb, &deleteHistory, cfg)
-			if err != nil {
-				return err
+		// round-robin to collect results from workers
+		for {
+			for i := 0; i < cfg.Workers; i++ {
+				select {
+				case <-stopChan:
+					return
+				case res, ok := <-workerOutputChannels[i]:
+					if !ok {
+						return
+					}
+
+					// filter out txs with no liveliness actions
+					if res.liveliness != nil && len(res.liveliness) > 0 {
+						orderedResults <- res
+					}
+				}
 			}
+		}
+	}()
+	return orderedResults
+}
+
+// launchWorkers lauches workers to process transactions in parallel.
+func launchWorkers(wg *sync.WaitGroup, cfg *utils.Config, workerInputChannels map[int]chan *substate.Transaction, processor executor.TxProcessor, stopChan chan struct{}, errChan chan error, log logger.Logger) map[int]chan txLivelinessResult {
```

Comment:
`launchWorker` is imo too general. How about `processSortedTx`?

Reply:
Fixed